h2o-r/h2o-package/R/explain.R

```diff
@@ -504,19 +504,28 @@ with_no_h2o_progress <- function(expr) {
     }
   }
 
-  for (feature in names(to_process)) {
-    col_parts <- strsplit(feature, ".", fixed = TRUE)[[1]]
-    found <- FALSE
-    for (prefix_len in seq(from = length(col_parts), to = 1, by = -1)){
-      prefix <- paste0(head(col_parts, n = prefix_len), collapse = ".")
-      if (prefix %in% x) {
-        consolidated_varimps[[prefix]] <- consolidated_varimps[[prefix]] + varimps[[feature]]
-        found <- TRUE
-        break
+  domains <- .h2o.__get_domain_mapping(model)
+  col_domain_mapping <- list()
+  for (col in names(domains)) {
+    if (!is.null(domains[[col]])) {
+      for (domain in c("missing(NA)", domains[[col]])) {
+        col_domain_mapping[[paste0(col, ".", domain)]] <- col
       }
     }
-    if (!found)
+  }
+
+  if (anyDuplicated(names(col_domain_mapping))) {
+    dups <- duplicated(names(col_domain_mapping)) | duplicated(names(col_domain_mapping), fromLast = TRUE)
+    stop("Ambiguous encoding of the column x category pairs: ", col_domain_mapping[dups])
```

Comment:
Unfortunately, this can happen:
```
library(h2o)
df <- data.frame(y=rep_len(1:5, 1000))
df$a <- as.factor(c("b.c", "d"))
df$a.b <- as.factor(letters[1:5])
df$y <- df$y * c(1, 10)

glm <- h2o.glm(y="y", training_frame = as.h2o(df))
h2o.varimp(glm)
```
=>
```
variable relative_importance scaled_importance   percentage
1    a.b.c        1.330686e+01      1.000000e+00 2.278348e-01
2      a.d        1.330686e+01      1.000000e+00 2.278348e-01
3    a.b.e        1.060822e+01      7.971997e-01 1.816298e-01
4    a.b.a        1.060822e+01      7.971997e-01 1.816298e-01
5    a.b.d        5.287787e+00      3.973731e-01 9.053541e-02
6    a.b.b        5.287787e+00      3.973731e-01 9.053541e-02
7    a.b.c        1.953338e-14      1.467919e-15 3.344429e-16
```

(see the first and the last line of the varimp)